Small Improvements

scymtym opened the pull request:
In preparations for the `map-similar` and `corrections` functions and also for (partially) generated (but don't worry) initial documentation.

scymtym committed:
Export LOOKUP and INSERT which name dictionary protocol functions

phoe commented on test/dictionary.lisp:
```diff
@@ -0,0 +1,24 @@
+(test dictionary.lookup
+  "Make sure that `lookup' behaves correctly for an empty dictionary."
+  (let ((dictionary (make-instance 'spell::dictionary)))
+    (signals error (spell:lookup "" dictionary))
```
Does it make sense for an empty word to explicitly signal an error? Does it make sense to alternatively consider an empty/whitespace-only word a part of every dictionary, just like the empty type being a subtype of every type?

Asking because someone doing spellchecking on series of words might not want to filter out empty subsequences before calling `lookup`.

scymtym replied:
One problem would be which concrete value to return when asked about `""` since there is no `word` for that in the dictionary. I think I would rather change the code so that the suggested behavior can be configured by users as they see fit:

```cl
(defmethod lookup ((string string) (dictionary dictionary))
  (let ((length (length string))
        (result '()))
    #+ (or) (unless (plusp length)
              (error "~@<Query string must not be empty.~@:>"))
    (node-lookup (lambda (word)
                   (push word result))
                 string length (contents dictionary))
    result))

(let ((d (make-instance 'dictionary))) ; empty
  (lookup "" d))
; => NIL

(let ((d (compact (make-instance 'dictionary)))) ; empty
  (lookup "" d))
; => NIL

(let ((d (make-instance 'dictionary)))
  (insert (make-word "" :noun "") "" d)
  (lookup "" d))
; => (#<EXPLICIT-BASE-NOUN "" number:ANY case:NIL gender:NIL {10042468D3}>)
```

What do you think?

scymtym replied:
Solved as discussed on IRC.

scymtym committed:
Handle LOOKUP of the empty word gracefully

This way, there are no special cases, neither for the empty word nor
the empty dictionary and LOOKUP always means "is this word in this
dictionary".

scymtym committed:
test: Add tests for methods which are related to the DICTIONARY class

scymtym committed:
Add BITFIELD-SLOTS method for BITFIELD-MIXIN

scymtym committed:
Fix PRINT-ITEMS method for WORD

Use slot names instead of bitfield-slot names since the latter are
only hints and can be wrong if multiple bitfield-slots get coalesced.

scymtym committed:
Rename generic functions in trie node protocols

Primarily to avoid collisions with planned additions but the new names
also seem better for possibly exporting the protocols later.

scymtym committed:
Add low-level function MAP-NODE-ENTRIES for traversing all entries

scymtym committed:
Add high-level function MAP-ENTRIES for traversing all entries

scymtym committed:
test: MAP-DICTIONARY-ENTRIES considers all dictionary text files

scymtym committed:
test: Add ENGLISH.COMPARE/MAP-ENTRIES
